Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+            performSubscribe(completion: completion)
+            
+        case let .subscribingStageOne(instanceWrapper, completions):
+            // Add the new completion handler to the `completions` of the .subscribing state
+            state = .subscribingStageOne(instanceWrapper: instanceWrapper,
+                                         completions: completions.appending(completion))
+            
+        case let .subscribingStageTwo(instanceWrapper, resumableSubscription, completions):
+            // Add the new completion handler to the `completions` of the .subscribing state
+            state = .subscribingStageTwo(instanceWrapper: instanceWrapper,
+                                         resumableSubscription: resumableSubscription,
+                                         completions: completions.appending(completion))
+            
+        case .subscribed:
+            // Immediately invoke success, we're already subcribed
+            completion(.success)
+            
+        }
+    }
+    
+    func unsubscribe() {
+        
+        switch state {
+            
+        case .notSubscribed:
+            // Nothing to do
+            ()
+            
+        case let .subscribingStageOne(_, completions):
+            
+            state = .notSubscribed
+            
+            let error = Self.unsubscribeCalledWhileSubscribingError
+            
+            delegate?.subscription(self, didReceiveError: error)
+            
+            for completion in completions {
+                completion(.failure(error))
+            }
+            
+        case let .subscribingStageTwo(_, resumableSubscription, completions):
+            
+            resumableSubscription.onOpen = nil
+            resumableSubscription.onError = nil
+            resumableSubscription.end()
+            
+            state = .notSubscribed
+            
+            let error = Self.unsubscribeCalledWhileSubscribingError
+            
+            delegate?.subscription(self, didReceiveError: error)
+            
+            for completion in completions {
+                completion(.failure(error))
+            }
+            
+        case let .subscribed(_, resumableSubscription):
+            
+            resumableSubscription.onOpen = nil
+            resumableSubscription.onError = nil
+            resumableSubscription.end()
+            
+            state = .notSubscribed
+        }
+    }
+    
+    // MARK: - Private
+    
+    private func performSubscribe(completion: @escaping SubscribeHandler) {
+        
+        let instanceWrapper = dependencies.instanceWrapperFactory.makeInstanceWrapper(forType: .subscription(subscriptionType))
+        
+        let requestPath = makeRequestPath(for: subscriptionType)
+        let requestOptions = PPRequestOptions(method: HTTPMethod.SUBSCRIBE.rawValue, path: requestPath)
+        
+        let onEvent: InstanceWrapper.OnEvent = { [weak self] _, _, jsonDataAsAny in
```

Comment:
When we execute code in closures passed to `PusherPlatform` we need to be extra careful about thread safety. Otherwise, race conditions that currently exist in `PusherPlatform` will propagate to `PusherChatkit`.

Reply:
Not sure if this is just a statement or if there is a specific issue with the code here?

Reply:
I just wanted to point out that we might encounter such problems in future due to existing race conditions in `PusherPlatform` and in such case we would need to revisit our integration. I am not trying to say that there is something wrong with the code above.